proj/hps_accel/gateware/math.py

```diff
@@ -0,0 +1,131 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nmigen import Cat, Mux, Signal, unsigned, signed
+from nmigen.hdl.dsl import Module
+from nmigen_cfu import InstructionBase, SimpleElaboratable
+
+from .constants import Constants
+from .stream import Endpoint, connect
+
+SRDHM_INPUT_LAYOUT = [
+    ('a', signed(32)),
+    ('b', signed(32)),
+]
+
+
+class SaturatingRoundingDoubleHighMul(SimpleElaboratable):
+    """Performs an SRDHM operation.
+
+    This is approximately a 32x32 mulitiplication where only the high 32
+    bits are returned. It actually returns double the value, which is
+    possible because both inputs are signed and so the raw result of the
+    multiplication is only a 63 bit number.
+
+    Only tested for
+
+    * 'a' in the range +/- (64 * 16 * 128) (i.e 18bits including sign).
+      This is the range required for processing 4x4 convs with depths
+      up to 64.
+    * 'b' in the range 0x4000_0000 to 0x7000_0000 as this appears to be
+      the range used the hps model.
+
+    Attributes
+    ----------
+
+    input: Stream(SRDHM_INPUT_LAYOUT), in
+      The calculation to perform.
+
+    output: Stream(signed(32)), out
+      The result.
+
+    This pipeline neither provides nor respects back pressure.
+    """
+    PIPELINE_CYCLES = 3
+
+    def __init__(self):
+        self.input = Endpoint(SRDHM_INPUT_LAYOUT)
+        self.output = Endpoint(signed(32))
+
+    def delay(self, m, cycles, sig):
+        sr = Signal(cycles)
+        m.d.sync += sr.eq(Cat(sig, sr[:-1]))
+        return sr[-1]
+
+    def elab(self, m):
+        # Input is always ready
+        m.d.comb += self.input.ready.eq(1)
+
+        # Output's valid is input's valid, but delayed
+        m.d.comb += self.output.valid.eq(
+            self.delay(m, self.PIPELINE_CYCLES, self.input.valid))
+
+        # Cycle 0: register inputs
+        a = self.input.payload.a
+        reg_a = Signal(signed(32))
+        reg_b = Signal(signed(32))
+        m.d.sync += reg_a.eq(Mux(a >= 0, a, -a))
+        m.d.sync += reg_b.eq(self.input.payload.b)
+
+        # Cycle 1: multiply to register
+        # both operands are positive, so result always positive
+        reg_ab = Signal(signed(63))
+        m.d.sync += reg_ab.eq(reg_a * reg_b)
+
+        # Cycle 2: nudge, take high bits and sign
+        positive_2 = self.delay(m, 2, a >= 0)
+        nudged = reg_ab + Mux(positive_2, (1 << 30), (1 << 30) - 1)
+        high_bits = Signal(signed(32))
+        m.d.comb += high_bits.eq(nudged[31:])
+        with_sign = Mux(positive_2, high_bits, -high_bits)
+        m.d.sync += self.output.payload.eq(with_sign)
+
+
+class MathInstruction(InstructionBase):
+    """An instruction used to perform specialist math operations.
+
+    This instruction is likely to be replaced with more comprehensive
+    gateware in future.
+
+    Sets a configuration register from in0.
```

Comment:
I don't grok this sentence in this context, is it actually a copy-paste error?

Reply:
Thanks. Yes. That is what happened. Fixed.